packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts

```diff
@@ -0,0 +1,494 @@
+    super(props);
+    this.protocol = Protocol.HTTP2;
+  }
+}
+
+/**
+ * Represents the properties needed to define a GRPC Listener for Virtual Gateway
+ */
+class GrpcGatewayListener extends VirtualGatewayListener {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Protocol the listener implements
+   */
+  protected protocol: Protocol = Protocol.GRPC;
+
+  constructor(props: HttpGatewayListenerProps = {}) {
+    super();
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+  }
+
+  /**
+   * Called when the GatewayListener type is initialized. Can be used to enforce
+   * mutual exclusivity
+   */
+  public bind(_scope: cdk.Construct): VirtualGatewayListenerConfig {
+    return {
+      listener: {
+        portMapping: {
+          port: this.port,
+          protocol: Protocol.GRPC,
+        },
+        healthCheck: renderHealthCheck(this.healthCheck, {
+          port: this.port,
+          protocol: Protocol.GRPC,
+        }),
+      },
+    };
+  }
+}
+
+/**
+ * Basic configuration properties for a VirtualGateway
+ */
+export interface VirtualGatewayBaseProps {
+  /**
+   * Name of the VirtualGateway
+   *
+   * @default - A name is automatically determined
+   */
+  readonly virtualGatewayName?: string;
+
+  /**
+   * Listeners for the VirtualGateway. Only one is supported.
+   *
+   * @default - Single HTTP listener on port 8080
+   */
+  readonly listeners?: VirtualGatewayListener[];
+
+  /**
+   * Access Logging Configuration for the VirtualGateway
+   *
+   * @default - no access logging
+   */
+  readonly accessLog?: AccessLog;
+}
+
+/**
+ * Properties used when creating a new VirtualGateway
+ */
+export interface VirtualGatewayProps extends VirtualGatewayBaseProps {
+  /**
+   * The mesh which the VirtualGateway belongs to
+   */
+  readonly mesh: IMesh;
+}
+
+abstract class VirtualGatewayBase extends cdk.Resource implements IVirtualGateway {
+  /**
+   * Name of the VirtualGateway
+   */
+  public abstract readonly virtualGatewayName: string;
+
+  /**
+   * The Amazon Resource Name (ARN) for the VirtualGateway
+   */
+  public abstract readonly virtualGatewayArn: string;
+
+  /**
+   * The name of the mesh which the VirtualGateway belongs to
+   */
+  public abstract readonly mesh: IMesh;
+
+  protected readonly listeners = new Array<VirtualGatewayListenerConfig>();
+  protected readonly routes = new Array<CfnGatewayRoute>();
+
+  /**
+   * Utility method to add a list of listeners to this VirtualGateway
+   */
+  public addListeners(listeners: VirtualGatewayListener[]) {
+    if (listeners.length + this.listeners.length > 1) {
+      throw new Error('VirtualGateway may have at most one listener');
```

Comment:
Wait...? So why do we have a method called `addListeners()`, if there can only be a single listener...? And why does `VirtualGateway` take a `listeners` property, instead of a single `listener`...?

Reply:
So, this is future proofing. We currently only accept a single listener but we intend to support multiple listeners in the future: https://github.com/aws/aws-app-mesh-roadmap/issues/120

Thinking about the `addListeners` method though, it is dead code. You cannot specify a Virtual Gateway without a listener, and since you cannot have more than a single listener, you can never use this method. I'll remove it! 👍